cluster/src/main/java/org/apache/iotdb/cluster/log/UnCommittedEntryManager.java

```diff
@@ -0,0 +1,194 @@
+     * Note that the caller should ensure appendingEntries[0].index <= entries[entries.size()-1].index + 1.
+     * Note that the caller should ensure not to truncate entries which have been committed.
+     *
+     * @param appendingEntries request entries
+     */
+    public void truncateAndAppend(List<Log> appendingEntries) {
+        long after = appendingEntries.get(0).getCurrLogIndex();
+        long len = after - offset;
+        if (len < 0) {
+            // the logs are being truncated to before our current offset portion, which is committed entries
+            // unconditional obedience to the leader's request. Maybe throw a exception here is better
+            offset = after;
+            entries = appendingEntries;
+            logger.error("The logs which first index is {} are going to truncate committed logs", after);
+        } else if (len == entries.size()) {
+            // after is the next index in the entries
+            // directly append
+            entries.addAll(appendingEntries);
+        } else {
+            // clear conflict entries
+            // then append
+            logger.info("truncate the entries after index {}", after);
+            int truncateIndex = (int) (after - offset);
+            if (truncateIndex < entries.size()) {
+                entries.subList(truncateIndex, entries.size()).clear();
+            }
+            entries.addAll(appendingEntries);
+        }
+    }
+
+    /**
+     * Pack entries from low through high - 1, just like slice (entries[low:high]).
+     * offset <= low <= high.
+     * Note that caller must ensure low <= high.
+     *
+     * @param low  request index low bound
+     * @param high request index upper bound
+     */
+    public List<Log> getEntries(long low, long high) {
+        if (low > high) {
+            logger.debug("invalid unCommittedEntryManager getEntries: parameter: low({}) > high({})", low, high);
+        }
+        long upper = offset + entries.size();
+        if (low > upper) {
+            // don't throw a exception to support
+            // getEntries(low, Integer.MAX_VALUE) if low is larger than lastIndex.
+            logger.info("unCommittedEntryManager getEntries[{},{}) out of bound : [{},{}] , return empty ArrayList", low, high, offset, upper);
+            return new ArrayList<>();
+        }
+        if (low < offset) {
+            logger.debug("unCommittedEntryManager getEntries[{},{}) out of bound : [{},{}]", low, high, offset, upper);
+            low = offset;
+        }
+        if (high > upper) {
+            logger.info("unCommittedEntryManager getEntries[{},{}) out of bound : [{},{}] , adjust parameter 'high' to {}", low, high, offset, upper, upper);
+            // don't throw a exception to support getEntries(low, Integer.MAX_VALUE).
+            high = upper;
+        }
+        return entries.subList((int) (low - offset), (int) (high - offset));
```

Comment:
If the caller change the subList, our manager will be impacted. You can use a new arrayList to return or add some comment at the method

Reply:
Actually,In RaftLogManager's getEntries implemetation,a new list is created to deeply copy these references.So it is safe to change the list outside RaftLogManager.
Thanks for reviewing!